web3/manager.py

```diff
@@ -33,37 +38,36 @@ def __init__(self, web3, providers, middlewares=None):
         self.providers = providers
 
     web3 = None
-    _middlewares = None
     _providers = None
 
     @property
     def middlewares(self):
-        return self._middlewares or tuple()
+        return tuple(self.middleware_stack)
 
     @middlewares.setter
-    def middlewares(self, value):
-        self._middlewares = tuple(value)
+    def middlewares(self, new_middlewares):
+        self.middleware_stack = NamedElementStack(new_middlewares)
         self._generate_request_functions()
+        self.middleware_stack.on_change(self._generate_request_functions)
```

Comment:
This makes me feel funny.  It's totally pragmatic and I could almost call it *elegant* but it feels like cramming javascript into python.  I don't love the alternative of just ensuring that `_generate_request_functions()` is called at all points where middlewares are modified, but.... I also don't love that this intertwines these two objects or that it feels a bit *magic*.

Proposed 3rd way to do this?

* Change `NamedElementStack` to be `MiddlewareStack`.
* Initialize it with the `provider` instances as well.
* Move the request function generation into the `MiddlewareStack` object itself, maybe as a  computed property which is cached based on the providers and middlewares in the stack.

Thoughts?

Reply:
> feels like cramming javascript into python

I'd say it's not so much a javascript pattern as a general object-oriented pattern. But if we can get rid of it, it's still a complexity win, so let's look at alternatives.

> 3rd way
> * Change NamedElementStack to be MiddlewareStack.
> * Initialize it with the provider instances as well.
> * Move the request function generation into the MiddlewareStack object itself, maybe as a computed property which is cached based on the providers and middlewares in the stack.

I'm interpreting this idea like:

```
ms = MiddlewareStack(middleware, providers)
...
class RequestManager:
    def _make_request(self, method, params):
        for provider in self.providers:
            request_func_for_provider = ms.request_func(self.web3, provider)
            try:
                return request_func_for_provider(method, params)
            except CannotHandleRequest:
                continue
```

It has the advantage of moving the change observation out of RequestManager, making it simpler. But MiddlewareStack gets a lot more responsibility of managing all the providers, rather than mimicking a basic collection.

That extra responsibility would add significant new lines of code, just pushing the problem to MiddlewareStack.

So option # 4:

Instead of pushing the request generation into MiddlewareStack, push it into the provider. Why is that any better? Each Provider only cares about its own middlewares, so no one is keeping a list of request functions or lists of MiddlewareStacks around, or regenerating caches for all providers at once.

A new `BaseProvider.request_func(web3, outer_middlewares)` would implement `combine_middlewares` so RequestManager can make a request with:

```
class RequestManager:
    def _make_request(self, method, params):
        for provider in self.providers:
            make_request_fn = provider.request_func(self.web3, self.middleware_stack)
            try:
                return make_request_fn(method, params)
            except CannotHandleRequest:
                continue
```

Then:

1. `RequestManager` never messes with provider middleware internals at all (encapsulation win)
2. Caching logic moves from `RequestManager` to `BaseProvider`. RequestManager was managing a list of cached request functions, BaseProvider is only caching a single request function (complexity win)
3. `on_change` becomes unnecessary (complexity win)
3. Don't have to generate the request function until the provider needs it (tiny performance win)
4.  Pushes a little extra logic into Providers, making testing harder for any provider that doesn't extend BaseProvider (testing loss)
5. Misbehaving Providers could theoretically ignore RequestManager middlewares... but why? (encapsulation loss)

Reply:
I'm +1 on moving it into the providers.  Seems the best option on the table that I've seen.